internal/config/cgmgr/cgmgr.go

```diff
@@ -159,3 +164,39 @@ func MoveProcessToContainerCgroup(containerPid, commandPid int) error {
 	}
 	return nil
 }
+
+func createSandboxCgroup(sbParentPath, containerCgroupName string) error {
+	cg := &cgcfgs.Cgroup{
+		Name:   containerCgroupName,
+		Parent: sbParentPath,
+		Resources: &cgcfgs.Resources{
+			SkipDevices: true,
+		},
+	}
+	mgr, err := libctrCgMgr.New(cg)
+	if err != nil {
+		return err
+	}
+
+	return mgr.Apply(-1)
```

Comment:
Well this does not make much sense to me.

1. Since you haven't specified `UseSystemd: true`, you are using fs manager. It just creates a bunch of directories.
2. It's a scope, so its lifetime is bound to the existence of at least one process in the scope. As we don't have any processes, systemd may just remove this.
3. I am not sure that having a `name=systemd` controller in there helps with (2), but overall this smells bad -- we create something that looks very much like what systemd would create, but systemd knows nothing about it.

Let's step back. Here is my limited understanding of the problem.

We have a pod (which is a scope in systemd terms), and a bunch of containers in it. We used to have a dummy infra container running a single pause process in order for cadvisor to identify the pod and get its stats. Now we don't have an infra container, and other containers may come and go, so there's no container that will always exist in the pod, and as a result cadvisor loses the pod (stops reporting its stats).

If what's written above is correct, than this is a problem with the approach of referring to a pod via a particular container which is in it. Containers can come and go,  and a pod should be identified by a pod cgroup path or something like that, not via any container.

Reply:
> so there's no container that will always exist in the pod, and as a result cadvisor loses the pod (stops reporting its stats)

Sort of. cadvisor assumes there's an infra container and an associated cgroup. the only way it registers network metrics is if there's a cgroup associated with the pod container. We can trick it to gather network metrics for the pod by pretending like there's the infra container by creating this dummy cgroup. I agree it smells bad--I've been wanting to change this interation for years with CRI stats KEP, but it's been delayed a number of times.

addressing the points more directly:
> Since you haven't specified UseSystemd: true, you are using fs manager. It just creates a bunch of directories.

yes this is the point. libcontainer will create the empty cgroup, and cadvisor (Which as inotify on the cgroup hierarchy) will catch its creation, see the pod ID, and then ask cri-o about that pod

> It's a scope, so its lifetime is bound to the existence of at least one process in the scope. As we don't have any processes, systemd may just remove this.

I was wondering whether this will happen. I believe this is the cause of the issue requiring https://github.com/cri-o/cri-o/pull/6196. I am not sure if this PR will help that or not, I am hoping it wil..

> I am not sure that having a name=systemd controller in there helps with (2), but overall this smells bad -- we create something that looks very much like what systemd would create, but systemd knows nothing about it.

agreed, though I'm not sure we can achieve anything better in the meantime. Folks need their network metrics, and we want systemd to not remove the cgroup. if we can achieve both of those internally in cri-o, it will be worth it for now, while we work on fixing it more legitimately